### PR TITLE
Ensure deploy key with write access can push

### LINKED ISSUE #19009: Deploy key with write access can not push
```
2022/03/06 00:00:42 router: completed GET /api/internal/serv/command/21/org/repo?mode=2&verb=git-receive-pack for 127.0.0.1:34602, 200 OK in 6.2ms @ private/serv.go:81(private.ServCommand)

2022/03/06 00:00:42 router: completed POST /api/internal/hook/pre-receive/org/repo for 127.0.0.1:34604, 403 Forbidden in 5.8ms @ private/hook_pre_receive.go:107(private.HookPreReceive)

2022/03/06 00:00:42 router: completed POST /api/internal/ssh/21/update/42 for 127.0.0.1:34606, 200 OK in 10.4ms @ private/key.go:18(private.UpdatePublicKeyInRepo)
```

The `HookPreReceive` calls `AssertCanWriteCode`, but `AssertCanWriteCode` only checks user's permission.

### PR DESCRIPTION
**Pull request by wxiaoguang**
Fix: 
*  #19009

In this PR, I use `DeployKeyID` to replace the `IsDeployKey`, then `CanWriteCode` can use the DeployKeyID to check the permission.

### CODE REVIEW

File: cmd/serv.go [hunk 1]
```diff
@@ -243,7 +243,7 @@ func runServ(c *cli.Context) error {
 	os.Setenv(models.EnvPusherID, strconv.FormatInt(results.UserID, 10))
 	os.Setenv(models.EnvRepoID, strconv.FormatInt(results.RepoID, 10))
 	os.Setenv(models.EnvPRID, fmt.Sprintf("%d", 0))
-	os.Setenv(models.EnvIsDeployKey, fmt.Sprintf("%t", results.IsDeployKey))
+	os.Setenv(models.EnvDeployKeyID, fmt.Sprintf("%d", results.DeployKeyID))
 	os.Setenv(models.EnvKeyID, fmt.Sprintf("%d", results.KeyID))
```
Comment (lunny) [inline cmd/serv.go:247]: Isn't `EnvKeyID` the deploy key id?

Comment (wxiaoguang) [reply cmd/serv.go:247]: The old code is messy. I had updated the comments in this PR.


image

image

image

File: routers/private/hook_pre_receive.go [hunk 2]
```diff
@@ -450,24 +444,42 @@ func generateGitEnv(opts *private.HookOptions) (env []string) {
...
+
+	user, err := user_model.GetUserByID(ctx.opts.UserID)
 	if err != nil {
-		log.Error("Unable to get User id %d Error: %v", id, err)
+		log.Error("Unable to get User id %d Error: %v", ctx.opts.UserID, err)
 		ctx.JSON(http.StatusInternalServerError, private.Response{
-			Err: fmt.Sprintf("Unable to get User id %d Error: %v", id, err),
+			Err: fmt.Sprintf("Unable to get User id %d Error: %v", ctx.opts.UserID, err),
 		})
 		return
```
Comment (Gusted) [inline routers/private/hook_pre_receive.go]: Should it return a error, so code wouldn't assume that `user` and userPerm` is non-nill?

Comment (wxiaoguang) [reply routers/private/hook_pre_receive.go]: Returning errors doesn't make code more simple at the moment.

The existing (old) logic is using `ctx.Written` to check errors in `loadPusherAndPermission`, the code in this PR just follows that and can guarantee that there is no nil access (and `userPerm` can never be nil)

If we want to rewrite the error handling logic, there would be a bigger refactoring, which makes the PR more complex and harder to be approved.

**Review by Gusted — APPROVED** [review]

File: routers/private/hook_pre_receive.go (hunk 2 above)
Comment (wxiaoguang) [reply routers/private/hook_pre_receive.go]: Hmm, I think I can make loadPusherAndPermission to return `bool`, then the logic doesn't change much, and can be more clear.

Comment (wxiaoguang) [reply routers/private/hook_pre_receive.go]: improved, pushed.

Comment (Gusted) [reply routers/private/hook_pre_receive.go]: Ack

File: routers/api/v1/repo/key.go [hunk 3]
```diff
@@ -144,7 +144,7 @@ func GetDeployKey(ctx *context.APIContext) {
 	//   "200":
 	//     "$ref": "#/responses/DeployKey"
 
-	key, err := asymkey_model.GetDeployKeyByID(db.DefaultContext, ctx.ParamsInt64(":id"))
+	key, err := asymkey_model.GetDeployKeyByID(ctx.ParamsInt64(":id"))
```
Comment (6543) [inline routers/api/v1/repo/key.go]:
```suggestion
	key, err := asymkey_model.GetDeployKeyByIDCtx(ctx, ctx.ParamsInt64(":id"))
```

File: models/asymkey/ssh_key_deploy.go [hunk 4]
```diff
@@ -167,7 +167,12 @@ func AddDeployKey(repoID int64, name, content string, readOnly bool) (*DeployKey
 }
 
 // GetDeployKeyByID returns deploy key by given ID.
-func GetDeployKeyByID(ctx context.Context, id int64) (*DeployKey, error) {
+func GetDeployKeyByID(id int64) (*DeployKey, error) {
```
Comment (6543) [inline models/asymkey/ssh_key_deploy.go]: we dont need that helper func just use the context you got

Comment (wxiaoguang) [reply models/asymkey/ssh_key_deploy.go]: Updated. Now there is only one `GetDeployKeyByID(ctx, id)`

**Review by 6543 — APPROVED** [review]

**Comment by wxiaoguang** [comment]
@6543  are you sure the `APIContext` can be used as `db.Context`?

image

**Comment by wxiaoguang** [comment]
There are a lot of code using `db.DefaultContext` in `routers`, maybe they should all be refactored to the new framework.

```
$ grep -r "db\.DefaultContext" ./routers | wc -l
56
```

**Comment by silverwind** [comment]
Backport necessary?

**Comment by zeripath** [comment]
I'm confused as to how this was failing because our testsuite should have picked this up... It would be helpful if you could add a test to our integrations suite that replicates the bug or give us a run down of the steps necessary to reproduce the original issue.

**Comment by wxiaoguang** [comment]
I have read the test cases before, if I understand correctly it seems that this case is not covered. I had the thought to write some tests at the same time however it's more complex than I thought, and I need this to be fixed in my server, so I just submitted the fix first. (although not fully covered by tests, but not too bad either)

I have a plan to do some further refactoring about the ssh/serv/internal system, and I will add more tests then.

About backport: this PR changes some env vars, it not as simple as a bug fix, so I think we can just keep it in 1.17. If anyone in 1.16 are affected, we can suggest them to use 1.17-dev or have further discussion then. Well ..... it seems that no one besides me need this feature 😂

**Comment by wxiaoguang** [comment]
About "the steps necessary to reproduce the original issue.":  that's quite straight, add a deploy key with write access, then you can not use the key to push. Because there was no code to check the key's permission to allow write.

**Pull request merged**